tests/components/image_processing/test_clarifai_general.py

```diff
@@ -0,0 +1,161 @@
+                                                    {'name': 'cat',
+                                                     'value': 0.14568}]}}]}
+
+PARSED_CONCEPTS = {'cat': 14.57, 'dog': 85.43}
+
+VALID_ENTITY_ID = 'image_processing.clarifai_demo_camera'
+VALID_CONFIG = {
+    ip.DOMAIN: {
+        'platform': 'clarifai_general',
+        cg.CONF_API_KEY: MOCK_API_KEY,
+        ip.CONF_SOURCE: {
+            ip.CONF_ENTITY_ID: 'camera.demo_camera'},
+    },
+    'camera': {
+        'platform': 'demo'
+        }
+    }
+
+
+class MockErrorResponse:
+    """Mock Clarifai RESPONSE to bad API key."""
+
+    status_code = 404
+    reason = 'Failure'
+    content = json.dumps({'status': {'description': 'API key not found'}})
+
+    @staticmethod
+    def json():
+        """Handle json."""
+        return {}
+
+
+RESOURCE = 'https://www.mock.com/url'
+PARAMS = {}
+METHOD = 'GET'
+RESPONSE = MockErrorResponse()
+ERROR = ApiError(RESOURCE, PARAMS, METHOD, RESPONSE)
+
+
+@pytest.fixture
+def mock_app():
+    """Return a mock ClarifaiApp object."""
+    with patch('clarifai.rest.ClarifaiApp') as _mock_app:
+        yield _mock_app
+
+
+@pytest.fixture
+def mock_app_with_error():
+    """Throw an ApiError."""
+    with patch('clarifai.rest.ClarifaiApp',
+               side_effect=ERROR) as _mock_mock_app_with_error:
+        yield _mock_mock_app_with_error
+
+
+@pytest.fixture
+def mock_image():
+    """Return a mock camera image."""
+    with patch('homeassistant.components.camera.demo.DemoCamera.camera_image',
+               return_value=b'Test') as image:
+        yield image
+
+
+def test_encode_image():
+    """Test that binary data is encoded correctly."""
+    assert cg.encode_image(b'test') == b'dGVzdA=='
+
+
+def test_parse_data():
+    """Test parsing of raw face data, and generation of matched_faces."""
+    raw_concepts = MOCK_RESPONSE['outputs'][0]['data']['concepts']
+    assert cg.parse_concepts(raw_concepts) == PARSED_CONCEPTS
+
+
+def test_valid_api_key(mock_app):
+    """Test that the api key is validated."""
+    cg.validate_api_key(MOCK_API_KEY)
+    assert mock_app.call_count == 1
+    assert mock_app.call_args == call(api_key=MOCK_API_KEY)
+
+
+def test_invalid_api_key(mock_app_with_error, caplog):
+    """Test that an invalid api key is caught."""
+    assert cg.validate_api_key(MOCK_API_KEY) is None
+    assert "Clarifai error: API key not found" in caplog.text
+
+
+async def test_setup_platform(hass, mock_app, mock_image):
+    """Set up platform with one entity."""
+    await async_setup_component(hass, ip.DOMAIN, VALID_CONFIG)
+    assert hass.states.get(VALID_ENTITY_ID)
+
+
+async def test_process_image(hass, mock_app, mock_image):
+    """Test successful processing of an image."""
+    await async_setup_component(hass, ip.DOMAIN, VALID_CONFIG)
+    assert hass.states.get(VALID_ENTITY_ID)
+
+    data = {ATTR_ENTITY_ID: VALID_ENTITY_ID}
+
+    with patch.object(
+        'clarifai.rest.client.Model', 'predict_by_base64', 
```

Comment:
@rok-povsic I am unsure how to patch `predict_by_base64` can you advise?

Reply:
@robmarkcole This is not connected with the Clarifai library itself but with monkey-patching and async. I suspect those two can't be used at the same time though I don't know because I'm not expert here. I tried to make it work but unfortunately can't spend more time here. If I were you I would either:
1. Try removing the `async` usage in tests if possible, similar to some other tests.
2. Not monkey-patch in the first place, but create a wrapper around the Clarifai library (using Abstract Base Classes) which would allow to create a fake implementation in a test, and inject it. I think this is a better approach in general, compared to monkey-patching. Though this might need a bigger refactoring of the project.

Reply:
Thanks for your advice @rok-povsic I have quite a full plate right now, @MartinHjelmare whats your thoughts on getting this merged with reduced test coverage (about 85% I think)?